speedtest-ookla-exporter: init at 1.1.4 + nixos/prometheus-exporters: init speedtest-ookla-exporter

emaiax opened the pull request:
This PR reintroduces the `prometheus-speedtest-ookla-exporter` package and NixOS module for configuring the Prometheus speedtest exporter, available as `services.prometheus.exporters.speedtest-ookla`. This enables monitoring of network speed metrics via Prometheus using the Ookla Speedtest CLI.

speedtest-exporter timeline:

- https://github.com/NixOS/nixpkgs/pull/203012, PR adding the previous speedtest-exporter
- https://github.com/NixOS/nixpkgs/issues/233062, discussions around changing the speedtest-exporter to [caarlos0/speedtest-exporter](https://github.com/caarlos0/speedtest-exporter), but closed as `not planned`
- https://github.com/NixOS/nixpkgs/pull/246320, package removal due to lack of maintainers

The [caarlos0/speedtest-exporter](https://github.com/caarlos0/speedtest-exporter) has been working for years without issues, so I think it's a solid option to bring this exporter back. This exporter requires [pkgs/by-name/oo/ookla-speedtest](https://github.com/NixOS/nixpkgs/blob/7905606cfa51a1815787377b9cb04291e87ebcb4/pkgs/by-name/oo/ookla-speedtest/package.nix) to be available in runtime; however, speedtest-cli is proprietary software.

**Naming note:** this was originally named `prometheus-speedtest-exporter` / `services.prometheus.exporters.speedtest`, but those names collided with a different, already-merged exporter ([podocarp/speedtest_exporter](https://github.com/podocarp/speedtest_exporter)) using the same attribute and option names. Renamed to `speedtest-ookla` to disambiguate, following the `<generic>-<distinguisher>` convention already used by `junos-czerwonk`.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [x] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/6134

acid-bong commented:
the correct title should be `nixos/PROGRAM`: https://github.com/NixOS/nixpkgs/blob/master/nixos/README.md#commit-conventions

acid-bong reviewed (changes requested):
actually no: split your PR into three commits:
1. new package (without test in passthru)
2. new module
3. new test (and add it to the package's passthru)

See common conventions ([one commit per logical unit](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions)), individual title conventions + follow dependency order. The final title format should be `PROGRAM: init at VERSION + nixos module`

emaiax committed:
nixos/prometheus: init speedtest-exporter

Introduces a new NixOS module for the Prometheus speedtest exporter, allowing configuration of debug mode and refresh interval.

emaiax committed:
nixos/prometheus: init speedtest-ookla-exporter

Introduces a new NixOS module for the Prometheus Ookla speedtest exporter, allowing configuration of debug mode and refresh interval. Named `speedtest-ookla` (not `speedtest`) since that option is already taken by an unrelated exporter (podocarp/speedtest_exporter).

acid-bong commented on pkgs/by-name/pr/prometheus-speedtest-exporter/package.nix:
```diff
@@ -0,0 +1,38 @@
+    description = "Exports speedtests as prometheus metrics";
+    mainProgram = "speedtest-exporter";
+    maintainers = with lib.maintainers; [ emaiax ];
+    license = lib.licenses.free;
```
the license is Apache
```suggestion
    license = lib.licenses.asl20;
```

emaiax commented on pkgs/by-name/pr/prometheus-speedtest-exporter/package.nix:
```diff
@@ -0,0 +1,38 @@
+    "-w"
+    "-X main.builtBy=nixpkgs"
+    "-X main.commit=unknown"
+    "-X main.date=unknown"
```
There are other compiled packages that I used as examples:

- https://github.com/emaiax/nixpkgs/blob/prometheus-speedtest-exporter/pkgs/by-name/pr/prometheus-alertmanager/package.nix
- https://github.com/emaiax/nixpkgs/blob/prometheus-speedtest-exporter/pkgs/by-name/pr/prometheus-mysqld-exporter/package.nix
- https://github.com/emaiax/nixpkgs/blob/prometheus-speedtest-exporter/pkgs/by-name/pr/prometheus-node-exporter/package.nix

emaiax committed:
nixos/prometheus: add speedtest-exporter tests

Add tests for the prometheus-speedtest-exporter, enabling service, debug mode, and refresh interval configuration. Also links the test in the package's passthru.tests for integration with NixOS test infrastructure.

emaiax committed:
nixos/prometheus: add speedtest-ookla-exporter tests

Add tests for the prometheus-speedtest-ookla-exporter, enabling service, debug mode, and refresh interval configuration. Also links the test in the package's passthru.tests for integration with NixOS test infrastructure.

emaiax committed:
speedtest-exporter: init at 1.1.4

Reintroduces `speedtest-exporter` enabling Prometheus metrics export for Ookla speedtests (runtime dependency).

emaiax committed:
speedtest-ookla-exporter: init at 1.1.4

Reintroduces `speedtest-exporter` (upstream caarlos0/speedtest-exporter) as `speedtest-ookla-exporter`, enabling Prometheus metrics export for Ookla speedtests (runtime dependency). Named with the `ookla` qualifier to avoid colliding with the unrelated `speedtest_exporter` (podocarp) already in nixpkgs.

acid-bong commented:
sry, burned out and can't review shit. good luck

emaiax committed:
nixos/prometheus: release note for speedtest-ookla-exporter

bjornfor commented on nixos/modules/services/monitoring/prometheus/exporters/speedtest-ookla.nix:
```diff
@@ -0,0 +1,64 @@
+    serviceConfig = {
+      StateDirectoryMode = "0750";
+      StateDirectory = "prometheus-speedtest-ookla-exporter";
+      WorkingDirectory = "/var/lib/prometheus-speedtest-ookla-exporter";
```
WorkingDirectory= is documented to be a relative directory (or have the special prefix `~`). I'm not sure what systemd does here, but it seems using `WorkingDirectory = "prometheus-speedtest-ookla-exporter"` is the most compatible option.

bjornfor commented on nixos/modules/services/monitoring/prometheus/exporters/speedtest-ookla.nix:
```diff
@@ -0,0 +1,64 @@
+
+      Environment = [
+        "PATH=${makeBinPath [ pkgs.ookla-speedtest ]}:$PATH"
+        "HOME=/var/lib/prometheus-speedtest-ookla-exporter"
```
FYI, I used to run this exporter with HOME set to a non-persistent directory, for better determinism. Never had an issue:

  serviceConfig.RuntimeDirectory = "prometheus-speedtest-ookla-exporter"
  environment.HOME = "/run/prometheus-speedtest-ookla-exporter"

bjornfor commented on nixos/modules/services/monitoring/prometheus/exporters/speedtest-ookla.nix:
```diff
@@ -0,0 +1,64 @@
+
+  serviceOpts = {
+    serviceConfig = {
+      StateDirectoryMode = "0750";
```
This seems to be needlessly running as root now; it can run as DynamicUser.
